pkg/proxy/iptables/number_generated_rules_test.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package iptables
+
+import (
+	"fmt"
+	"net"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/component-base/metrics/testutil"
+	"k8s.io/kubernetes/pkg/proxy/metrics"
+	utiliptables "k8s.io/kubernetes/pkg/util/iptables"
+	iptablestest "k8s.io/kubernetes/pkg/util/iptables/testing"
+	netutils "k8s.io/utils/net"
+	utilpointer "k8s.io/utils/pointer"
+)
+
+// kube-proxy generates iptables rules to forward traffic from Services to Endpoints
+// kube-proxy uses iptables-restore to configure the rules atomically, however,
+// this has the downside that large number of rules take a long time to be processed,
+// causing disruption.
+// There are different parameters than influence the number of rules generated:
+// - ServiceType
+// - Number of Services
+// - Number of Endpoints per Service
+// This test will fail when the number of rules change, so the person
+// that is modifying the code can have feedback about the performance impact
+// on their changes. It also runs multiple number of rules test cases to check
+// if the number of rules grows linearly.
+func TestNumberIptablesRules(t *testing.T) {
+	testCases := []struct {
+		name                string
+		serviceType         v1.ServiceType
+		services            int
+		epPerService        int
+		expectedNatRules    int
+		expectedFilterRules int
+	}{
+		{
+			name:                "ClusterIP 500 Services 500 EndpointPerService",
+			serviceType:         v1.ServiceTypeClusterIP,
+			services:            500,
+			epPerService:        500,
+			expectedNatRules:    2505,
```

Comment:
These numbers do not add up.

I patched it in, changed 500 to 5, and what I see is that it only generates 1 EP per service.  I did not debug it for you, sorry:

```
$ go test -v ./pkg/proxy/iptables/ -run TestNumberIptablesRules
=== RUN   TestNumberIptablesRules
=== RUN   TestNumberIptablesRules/ClusterIP_500_Services_500_EndpointPerService
E0610 12:12:03.572836 2591205 proxier.go:1449] TIM: -A KUBE-POSTROUTING -m mark ! --mark 0x4000/0x4000 -j RETURN
-A KUBE-POSTROUTING -j MARK --xor-mark 0x4000
-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE
-A KUBE-MARK-MASQ -j MARK --or-mark 0x4000
-A KUBE-SEP-CEKR4MZD3PGPVZWT -m comment --comment namespace/svc0:8080 -s 172.16.0.1 -j KUBE-MARK-MASQ
-A KUBE-SEP-CEKR4MZD3PGPVZWT -m comment --comment namespace/svc0:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.1:8080
-A KUBE-SVC-7S3BOWXDKQS3LONZ -m comment --comment "namespace/svc0:8080 cluster IP" -m tcp -p tcp -d 10.0.0.1 --dport 80 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m comment --comment "namespace/svc0:8080 cluster IP" -m tcp -p tcp -d 10.0.0.1 --dport 80 -j KUBE-SVC-7S3BOWXDKQS3LONZ
-A KUBE-SVC-7S3BOWXDKQS3LONZ -m comment --comment "namespace/svc0:8080 -> 172.16.0.1:8080" -j KUBE-SEP-CEKR4MZD3PGPVZWT
-A KUBE-SEP-AGR2HYPNMUY42VHO -m comment --comment namespace/svc1:8080 -s 172.16.0.1 -j KUBE-MARK-MASQ
-A KUBE-SEP-AGR2HYPNMUY42VHO -m comment --comment namespace/svc1:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.1:8080
-A KUBE-SVC-YLKEUCZCZYIKLXCJ -m comment --comment "namespace/svc1:8080 cluster IP" -m tcp -p tcp -d 10.0.0.2 --dport 81 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m comment --comment "namespace/svc1:8080 cluster IP" -m tcp -p tcp -d 10.0.0.2 --dport 81 -j KUBE-SVC-YLKEUCZCZYIKLXCJ
-A KUBE-SVC-YLKEUCZCZYIKLXCJ -m comment --comment "namespace/svc1:8080 -> 172.16.0.1:8080" -j KUBE-SEP-AGR2HYPNMUY42VHO
-A KUBE-SEP-YFZLR7PNQZ57O5LO -m comment --comment namespace/svc2:8080 -s 172.16.0.1 -j KUBE-MARK-MASQ
-A KUBE-SEP-YFZLR7PNQZ57O5LO -m comment --comment namespace/svc2:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.1:8080
-A KUBE-SVC-WPRXGYIHEJJLFXZB -m comment --comment "namespace/svc2:8080 cluster IP" -m tcp -p tcp -d 10.0.0.3 --dport 82 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m comment --comment "namespace/svc2:8080 cluster IP" -m tcp -p tcp -d 10.0.0.3 --dport 82 -j KUBE-SVC-WPRXGYIHEJJLFXZB
-A KUBE-SVC-WPRXGYIHEJJLFXZB -m comment --comment "namespace/svc2:8080 -> 172.16.0.1:8080" -j KUBE-SEP-YFZLR7PNQZ57O5LO
-A KUBE-SEP-7Z7RIUNVMXASXJYK -m comment --comment namespace/svc3:8080 -s 172.16.0.1 -j KUBE-MARK-MASQ
-A KUBE-SEP-7Z7RIUNVMXASXJYK -m comment --comment namespace/svc3:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.1:8080
-A KUBE-SVC-GXKQ24MNRZGPA63O -m comment --comment "namespace/svc3:8080 cluster IP" -m tcp -p tcp -d 10.0.0.4 --dport 83 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m comment --comment "namespace/svc3:8080 cluster IP" -m tcp -p tcp -d 10.0.0.4 --dport 83 -j KUBE-SVC-GXKQ24MNRZGPA63O
-A KUBE-SVC-GXKQ24MNRZGPA63O -m comment --comment "namespace/svc3:8080 -> 172.16.0.1:8080" -j KUBE-SEP-7Z7RIUNVMXASXJYK
-A KUBE-SEP-WIOLMW6VRD2EW6AE -m comment --comment namespace/svc4:8080 -s 172.16.0.1 -j KUBE-MARK-MASQ
-A KUBE-SEP-WIOLMW6VRD2EW6AE -m comment --comment namespace/svc4:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.1:8080
-A KUBE-SVC-YTLTP7CASMWZTGMX -m comment --comment "namespace/svc4:8080 cluster IP" -m tcp -p tcp -d 10.0.0.5 --dport 84 ! -s 10.0.0.0/8 -j KUBE-MARK-MASQ
-A KUBE-SERVICES -m comment --comment "namespace/svc4:8080 cluster IP" -m tcp -p tcp -d 10.0.0.5 --dport 84 -j KUBE-SVC-YTLTP7CASMWZTGMX
-A KUBE-SVC-YTLTP7CASMWZTGMX -m comment --comment "namespace/svc4:8080 -> 172.16.0.1:8080" -j KUBE-SEP-WIOLMW6VRD2EW6AE
-A KUBE-SERVICES -m comment --comment "kubernetes service nodeports; NOTE: this must be the last rule in this chain" -m addrtype --dst-type LOCAL -j KUBE-NODEPORTS
```

Reply:
> it only generates 1 EP per service

lol, https://github.com/kubernetes/kubernetes/issues/106267

Reply:
:facepalm:

Reply:
fixed
```
        -A KUBE-SEP-R3OJ5E3NOGQ5NY7C -m comment --comment namespace/svc7:8080 -s 172.16.0.5 -j KUBE-MARK-MASQ
        -A KUBE-SEP-R3OJ5E3NOGQ5NY7C -m comment --comment namespace/svc7:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.5:8080
        -A KUBE-SEP-ORO3SXPG5QTEBZBG -m comment --comment namespace/svc7:8080 -s 172.16.0.6 -j KUBE-MARK-MASQ
        -A KUBE-SEP-ORO3SXPG5QTEBZBG -m comment --comment namespace/svc7:8080 -m tcp -p tcp -j DNAT --to-destination 172.16.0.6:8080
        -A KUBE-SEP-GLLVMAHR576DMKU7 -m comment --comment namespace/svc7:8080 -s 172.16.0.7 -j KUBE-MARK-MASQ
```